### ISSUE TITLE
join and joinUnbounded release resources too early

### ISSUE THREAD

**Issue by mpilquist** [functional-streams-for-scala/fs2#967]
Consider:

```scala
Stream.bracket(IO(println("acquired")))(_ => Stream.emit(()), _ => IO(println("released"))).
  map(_ => Stream.eval(IO(Thread.sleep(500)) *> IO(println("using")))).
  joinUnbounded.run
```

Running this program results in:
```
acquired
released
using
```

Replacing `joinUnbounded` with `flatMap(identity)` results in the correct order.

Originally spotted by @tpolecat in `SocketSpec` where replacing `Stream.emit(s.flatMap(f))` with `s.map(f)` resulted in errors indicating the channel had been closed.

**Comment by mpilquist** [comment]
@pchlupacek Seems like the `runOuter` method in `join` is missing an `onFinalizer` that blocks on the inners completing. WDYT?

**Comment by pchlupacek** [comment]
@mpilquist take a look on that. I am just not 100% sure, but I recall fixing this in 0.9.x branch....

**Comment by mpilquist** [comment]
I added a failing test case in #968. I spot checked 0.9 and it appears to have the same issue.

**Comment by mpilquist** [comment]
Given `s: Stream[F,A]` and `f: A => Stream[F,B]`, resources allocated in scope of a single `A` must not be released until the inner `Stream[F,B]` completes. There are at least 2 cases to consider here: a resource allocated in process of generating a single `A` and resources allocated in generation of some/all `A`s in `s`. A client socket is an example of the first case whereas the bind socket is an example of the second case. The implementation should preserve the property that resources are freed eagerly (e.g., in the TCP example, releasing client sockets shouldn't happen at the end of the outer stream but rather as each inner stream completes).

I don't know if the current join implementation can be adapted to meet these requirements without some new APIs that let it manipulate scopes.

**Comment by pchlupacek** [comment]
@mpilquist Thanks & I'll take a look on this. I have feeling that resource thing can be resolved by prohibiting the outer stream to terminate before inner streams will. 

I just need to think in context of interruption what that would that mean, and if we wouldn't end up wit some deadlock.

**Comment by mpilquist** [comment]
@pchlupacek I have a prototype of a fix for this issue which involves adding a reference count to each resource and adding 2 new cases to the stream algebra -- one that exports all resources, incrementing their reference counts and another that imports all resources in to the current scope. The test case passes with this change (and with no other synchronization added to join).

**Comment by mpilquist** [comment]
@pchlupacek I cleaned up the prototype and pushed it to https://github.com/functional-streams-for-scala/fs2/pull/968.

**Comment by pchlupacek** [comment]
@mpilquist Thank you will take a look on this. I think still perhaps in addition to what you suggest I shall correct the join behaviour where the outer stream never terminates before the last inner stream does, which is not the current case.

**Comment by pchlupacek** [comment]
I mean finalizer's won't be run until the last inner stream is running.

**Comment by mpilquist** [comment]
Fixed by #968
